### PR TITLE
contract: implement Kip160 (TreasuryRebalanceV2)

### PR DESCRIPTION
**Pull request by yoomee1313**
## Proposed changes

- This PR implements Kip160 (TreasuryRebalanceV2). https://github.com/klaytn/kips/pull/160
- KIP160 contract address needs to be put in Kip160ContractAddress
```
"config": {
        ...
        "kip160CompatibleBlock": 200,
        "kip160ContractAddress": "...",
}
```

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [ ] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [ ] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [ ] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

Tested with below genesis.json
```
"kip103CompatibleBlock": 200,
"kip103ContractAddress": "0x2C5811e9A04b3a14cB2d7E620FfCf8936E763b68",
"kip160CompatibleBlock": 400,
"kip160ContractAddress": "0x...",
```
- [1] kip103 test -> another node's syncing succeed and memo follows kip103 naming rule
  - deploy node with `dev` branch
  - activate kip103 hardfork
  - deploy another node with `PR` branch for syncing
  - memo
```
INFO[03/27,07:01:27 +08] [24] successfully executed treasury rebalancing memo="{\"retired\":{\"0x140bd2d00402e1c6d34db83efa17e22dc213f8f2\":5000000000000000000,\"0x742986e9b527778227b64a58247523cdbbd60869\":4997854300000000000,\"0xfa5078f97240ddb421d0ddbe51900ebdef5537db\":5000000000000000000},\"newbie\":{\"0x8be5f418b882ab9b85690cf50939521da1e7e09f\":2000000000000000000,\"0x96a9b8e8731c80ee84196cfdb3ee943353968bd0\":4000000000000000000,\"0xb379d31c54e7489773c6fa3775c9b4860039614f\":3000000000000000000},\"burnt\":20997854300000000000,\"success\":true}"
```
- [2] kip160 test -> kip160 rebalancing succeed and memo follows kip160 naming rule
  - deploy node with 'PR` branch and configure kip160CompatibleBlock & Kip160ContractAddress
  - activate kip160 hardfork
  - checked balance of reserves are correctly rebalanced

### CODE REVIEW

**Review by hyunsooda — DISMISSED** [review]

**Comment by yoomee1313** [comment]
This PR has been rebased. Please take a look. @hyunsooda

**Comment by blukat29** [comment]
You can't add KIP160 address to the RandaoRegistry after Randao hardfork. ChainConfig.RandaoRegistry is inserted exactly once at the Randao fork block. You have to either (1) insert KIP160 address to the system contract Registry (at address 0x401) via transactions or (2) declare KIP160 address in ChainConfig.KIP160Address variable.

**Comment by yoomee1313** [comment]
Declared Kip160CompatibleBlock and kip160ContractAddress field in chainConfig and use it as an optional fork applicable only for kip160. https://github.com/klaytn/klaytn/pull/2122/commits/0470c04a8bedf1e3434fa0c205b6f3e6a03c8b2c, https://github.com/klaytn/klaytn/pull/2122/commits/123e1170651ffc7aec1767461142057a5112d396.

**Comment by hyunsooda** [comment]
@yoomee1313 Does it not need to fill empty values in Cypress and Baobab configuration structures?

File: blockchain/system/rebalance.go [hunk 1]
```diff
@@ -170,24 +263,24 @@ func RebalanceTreasury(state *state.StateDB, chain backends.BlockChainForCaller,
 		return result, ErrRebalanceBadStatus
 	}
 
-	// Validation 3) Check approvals from retirees
-	if err := caller.CheckRetiredsApproved(nil); err != nil {
+	// Validation  3) Check approvals from zeroeds
```
Comment (hyeonLewis) [inline blockchain/system/rebalance.go]:
```suggestion
	// Validation 3) Check approvals from zeroeds
```

File: blockchain/system/rebalance.go [hunk 2]
```diff
@@ -68,95 +130,126 @@ func (caller *Kip103ContractCaller) CallContract(ctx context.Context, call klayt
...
+		isKIP160 = chain.Config().IsKIP160ForkBlock(header.Number)
+		isKIP103 = chain.Config().IsKIP103ForkBlock(header.Number)
 	)
 
-	caller, err := rebalance.NewTreasuryRebalanceCaller(chain.Config().Kip103ContractAddress, c)
+	if isKIP160 {
+		caller, err = rebalance.NewTreasuryRebalanceV2Caller(chain.Config().Kip160ContractAddress, backends.NewBlockchainContractBackend(chain, nil, nil))
+	} else if isKIP103 {
+		caller, err = NewKip103ContractCaller(state, chain, header)
+	}
```
Comment (hyeonLewis) [inline blockchain/system/rebalance.go]:
```suggestion
	} else if isKIP103 {
		caller, err = NewKip103ContractCaller(state, chain, header)
	} else {
	        return nil, ErrRebalanceIncorrectBlock
	}
```

File: blockchain/system/rebalance.go (hunk 1 above)
Comment (yoomee1313) [reply blockchain/system/rebalance.go]: reflected https://github.com/klaytn/klaytn/pull/2122/commits/bbb4e03fbc5106ead36f6f01dab0abc085bfffe0

File: blockchain/system/rebalance.go (hunk 2 above)
Comment (yoomee1313) [reply blockchain/system/rebalance.go]: `ErrRebalanceIncorrectBlock` indicates the incorrect behavior of the contract, so instead, it returned the custom error message. Reflected at https://github.com/klaytn/klaytn/pull/2122/commits/bbb4e03fbc5106ead36f6f01dab0abc085bfffe0.

**Comment by yoomee1313** [comment]
> @yoomee1313 Does it not need to fill empty values in Cypress and Baobab configuration structures?

@hyunsooda No need to fill. It's okay. It will be updated after block number and contract address is decided.

**Comment by yoomee1313** [comment]
> @yoomee1313 Does it not need to fill empty values in Cypress and Baobab configuration structures?

**Comment by yoomee1313** [comment]
Seperated the two rebalance contracts to be located in seperate files and revised V2 contracts. commit: https://github.com/klaytn/klaytn/pull/2122/commits/e7fb74501cb86571d73ca6a1a54245c2dcf87561. cc. @blukat29

**Review by blukat29 — APPROVED** [review]

**Review by hyeonLewis — APPROVED** [review]

**Pull request merged**